File: src/main/java/com/machfour/macros/cli/Portion.java
package com.machfour.macros.cli;

import com.machfour.macros.linux.Config;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.objects.Food;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDatabase;
import com.machfour.macros.util.FoodPortionSpec;
import com.machfour.macros.util.PrintFormatting;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import static com.machfour.macros.cli.CliMain.OUT;
import static com.machfour.macros.cli.CliMain.PROGNAME;

class Portion extends CommandImpl {
    private static final String NAME = "portion";
    private static final String USAGE =
            String.format("Usage: %s %s [ <meal name> [<day>] -s ] <portion spec> [<portion spec> ... ]", PROGNAME, NAME);

    Portion() {
        super(NAME, USAGE);
    }

    @Override
    public void doAction(List<String> args) {
        if (args.size() == 1 || args.contains("--help")) {
            printHelp();
            return;
        }
        MacrosDatabase db = LinuxDatabase.getInstance(Config.DB_LOCATION);
        // MealSpec mealSpec = MealSpec.makeMealSpec(args);

        // check position of -s to decide how to parse arguments
        MealSpec mealSpec;
        int separator = args.indexOf("-s");
        switch (separator) {
            case 1:
                // it's the first argument after 'portion'
                // so basically there's no meal info anyway
                // fall-through to no flag
            case -1:
                // no meal info specified
                mealSpec = MealSpec.makeMealSpec();
                break;
            case 2:
                // args looks like ["portion", "<meal name>", "-s", ...]
                mealSpec = MealSpec.makeMealSpec(args.get(1));
                break;
            case 3:
                // args looks like ["portion", "<meal name>", "<day>", "-s", ...]
                mealSpec = MealSpec.makeMealSpec(args.get(1), args.get(2));
                break;
            case 0:
                assert false : "'-s' is where the command name should be!";
            default:
                out.println("There can only be at most two arguments before '-s'");
                return;
        }

        if (!mealSpec.mealSpecified()) {
            out.printf("No meal specified, assuming %s on %s\n", mealSpec.name(), PrintFormatting.prettyDay(mealSpec.day()));
        }
        mealSpec.process(db, true);
        if (mealSpec.error() != null) {
            out.println(mealSpec.error());
            return;
        }

        // now read everything after the '-s' as food portion specs
        List<FoodPortionSpec> specs = new ArrayList<>(args.size() - 1 - separator);
        for (int index = separator + 1; index < args.size(); ++index) {
            specs.add(FileParser.makefoodPortionSpecFromLine(args.get(index)));
        }

        process(mealSpec.processedObject(), specs, db);

    }

    static void process(Meal toAddTo, List<FoodPortionSpec> specs, MacrosDatabase db) {
        if (specs.isEmpty()) {
            out.println("No food portions specified, nothing to do");
            return;
        }

        FoodPortionSpec spec = specs.get(0);

        Food f;

        try {
            f = db.getFoodByIndexName(spec.foodIndexName);
        } catch (SQLException e) {
            out.println("Could not retrieve food. Reason: " + e.getMessage());
            return;
        }
        if (f == null) {
            out.printf("Unrecognised food with index name '%s'\n", spec.foodIndexName);
            return;
        }
        FileParser.processFpSpec(spec, toAddTo, f);
        if (spec.error != null) {
            out.println(spec.error);
            return;
        }
        assert (spec.createdObject != null) : "No object created but no error message either";
        toAddTo.addFoodPortion(spec.createdObject);

        try {
            db.saveFoodPortions(toAddTo);
        } catch (SQLException e) {
            out.println("Error saving food portion. Reason: " + e.getMessage());
            return;
        }
        out.println();
        MealPrinter.printMeal(toAddTo, false, OUT);

    }
}

File: macros-cli/src/main/kotlin/com/machfour/macros/cli/modes/Portion.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CliConfig
import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.makeMealSpec
import com.machfour.macros.cli.utils.printMeal
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.core.ObjectSource
import com.machfour.macros.entities.Food
import com.machfour.macros.entities.Meal
import com.machfour.macros.parsing.FileParser.Companion.makeFoodPortionSpecFromLine
import com.machfour.macros.parsing.FileParser.Companion.processFpSpec
import com.machfour.macros.parsing.FoodPortionSpec
import com.machfour.macros.parsing.MealSpec
import com.machfour.macros.queries.getFoodByIndexName
import com.machfour.macros.queries.saveObject
import com.machfour.macros.sql.SqlDatabase
import com.machfour.macros.sql.SqlException

fun List<FoodPortionSpec>.processAndAddTo(meal: Meal, db: SqlDatabase): Int {
    if (isEmpty()) {
        println("No food portions specified, nothing to do")
        return 0
    }
    for (spec in this) {
        val f: Food? = try {
            getFoodByIndexName(db, spec.foodIndexName)
        } catch (e: SqlException) {
            printlnErr("Could not retrieve food. Reason: " + e.message)
            return 1
        }
        if (f == null) {
            printlnErr("Unrecognised food with index name '${spec.foodIndexName}'")
            return 1
        }
        processFpSpec(spec, meal, f)
        if (spec.error.isNotEmpty()) {
            printlnErr(spec.error)
            return 1
        }
        val createdObject = checkNotNull(spec.createdObject) {
            "No object created but no error message either"
        }
        meal.addFoodPortion(createdObject)
    }
    try {
        meal.foodPortions
            .filter { it.source != ObjectSource.DATABASE }
            .forEach { saveObject(db, it) }
    } catch (e: SqlException) {
        printlnErr("Error saving food portion. Reason: " + e.message)
        return 1
    }
    println()
    printMeal(meal, false)
    return 0
}


class Portion(config: CliConfig): CommandImpl(config) {
    override val name = "portion"
    override val usage = "Usage: ${config.programName} $name [ <meal name> [<day>] -s ] <portion spec> [<portion spec> ... ]"

    override fun doAction(args: List<String>): Int {
        if (args.size == 1 || args.contains("--help")) {
            printHelp()
            return 0
        }
        val ds = config.database
        // MealSpec mealSpec = MealSpec.makeMealSpec(args);

        // TODO use argParsing here
        // check position of -s to decide how to parse arguments
        val mealSpec: MealSpec
        val separator = args.indexOf("-s")
        mealSpec = when (separator) {
            1, -1 -> makeMealSpec() // no meal info specified
            2 -> makeMealSpec(args[1]) // args looks like ["portion", "<meal name>", "-s", ...]
            3 -> makeMealSpec(args[1], args[2]) // args looks like ["portion", "<meal name>", "<day>", "-s", ...]
            0 -> {
                println("There can only be at most two arguments before '-s'")
                return 1
            }
            else -> {
                println("There can only be at most two arguments before '-s'")
                return 1
            }
        }
        if (!mealSpec.isMealSpecified) {
            val name = mealSpec.name
            val dayString = mealSpec.day?.prettyPrint() ?: "(invalid day)"
            println("No meal specified, assuming $name on $dayString")
        }
        mealSpec.process(ds, true)
        if (mealSpec.error != null) {
            printlnErr(mealSpec.error)
            return 1
        }

        // now read everything after the '-s' as food portion specs
        val specs = args.drop(separator + 1).map{ makeFoodPortionSpecFromLine(it) }
        val processedMeal = checkNotNull(mealSpec.processedObject) {
            "mealspec did not correctly process object but no error was given"
        }

        return specs.processAndAddTo(processedMeal, ds)
    }

}